File: topological_sorting/go/adjacency_list.go
package topological_sorting

type Graph struct {
	VertexCount int
	Directed    bool
	Edges       []Edge
}

type Edge struct {
	X int
	Y int
}

type AdjacencyList [][]int

type VertexCallback func(x int)

func NewAdjacencyList(g Graph) AdjacencyList {
	a := make(AdjacencyList, g.VertexCount)
	for _, edge := range g.Edges {
		insertEdge(a, edge.X, edge.Y, g.Directed)
	}
	return a
}

func insertEdge(a AdjacencyList, x, y int, directed bool) {
	a[x] = append(a[x], y)
	if !directed {
		insertEdge(a, y, x, true)
	}
}

func (a AdjacencyList) DepthFirstSearch(x int, lateCallback VertexCallback) {
	discovered := make([]bool, len(a))
	dfs(a, discovered, x, lateCallback)
}

func dfs(a AdjacencyList, discovered []bool, x int, lateCallback VertexCallback) {
	if discovered[x] {
		return
	}
	discovered[x] = true
	for _, y := range a[x] {
		dfs(a, discovered, y, lateCallback)
	}
	lateCallback(x)
}

File: dag_shortest_path/typescript/adjacency_list.ts
export interface Graph {
    vertexCount: number;
    directed: boolean;
    edges: Edge[];
}

export interface Edge {
    x: number;
    y: number;
    weight: number;
}

type VertexCallback = (x: number) => void;

export class AdjacencyList {
    a: Edge[][];
    constructor(graph: Graph) {
        this.a = constructAdjacencyList(graph);
    }
    depthFirstSearch(start: number, lateCallback: VertexCallback) {
        depthFirstSearch(this.a, start, lateCallback);
    }
}

function constructAdjacencyList(graph: Graph): Edge[][] {
    // new Array(vertexCount).fill([]) does not work because it reuses the same array instance
    // for every element.
    const a = Array.from(new Array(graph.vertexCount), () => []);

    graph.edges.forEach(edge => {
        insertEdge(a, edge, graph.directed);
    });

    return a;
}

function insertEdge(a: Edge[][], edge: Edge, directed: boolean): void {
    a[edge.x].push(edge);
    if (!directed) {
        const reversed = {x: edge.y, y: edge.x, weight: edge.weight};
        insertEdge(a, reversed, true);
    }
}

function depthFirstSearch(a: Edge[][], start: number, lateCallback: VertexCallback): void {
    const discovered = new Array(a.length);

    function dfs(x: number) {
        if (discovered[x]) {
            return;
        }
        discovered[x] = true;
        a[x].forEach(edge => dfs(edge.y));
        lateCallback(x);
    }

    dfs(start);
}
